File: src/skills/src/sDefendPoint.cpp
#include "skillSet.h"
#include <ssl_common/grSimComm.h>
#include <ssl_common/config.h>
#include <navigation/planners.h>
#include <navigation/controllers/waypoint.h>

#define DEFEND_RADIUS 50.0
using namespace std;
//addCircle..
namespace Strategy
{
	gr_Robot_Command SkillSet::defendPoint(const SParam &param, const BeliefState &state, int botID)
	{
		Vector2D<int> botPos(state.homePos[botID].x, state.homePos[botID].y);
		float point_ball_angle = atan2(state.ballPos.y-param.DefendPointP.y,state.ballPos.x-param.DefendPointP.x);
		Vector2D<int> dpoint;
		//comm.addCircle(param.DefendPointP.x,param.DefendPointP.y,50);
		dpoint.y=param.DefendPointP.y + (param.DefendPointP.radius)*sin(point_ball_angle);

		dpoint.x=param.DefendPointP.x + (param.DefendPointP.radius)*cos(point_ball_angle);


		using Navigation::obstacle;
	    vector<obstacle> obs;
	    obstacle o;
	    for (int i = 0; i < state.homeDetected.size(); ++i)
	    {
	      o.x = state.homePos[i].x;
	      o.y = state.homePos[i].y;
	      o.radius = 3 * BOT_RADIUS;
	      obs.push_back(o);
	    }

		for (int i = state.homeDetected.size(); i < state.homeDetected.size() + state.awayDetected.size(); ++i)
	    {
	      o.x = state.awayPos[i - state.homeDetected.size()].x;
	      o.y = state.awayPos[i - state.homeDetected.size()].y;
	      o.radius = 3 * BOT_RADIUS;
	      obs.push_back(o);
	    }
		Vector2D<int> point, nextWP, nextNWP;

		// use the mergescurver planner
   		Navigation::MergeSCurve pathPlanner;
		pathPlanner.plan(botPos,
		                  dpoint,
		                  &nextWP,
		                  &nextNWP,
		                  obs,
		                  state.homeDetected.size() + state.awayDetected.size(),
		                  botID,
		                  true);

		if(nextWP.valid())
		{
		// 	comm.addCircle(nextWP.x, nextWP.y, 50);
		// 	comm.addLine(state.homePos[botID].x, state.homePos[botID].y, nextWP.x, nextWP.y);
		}
		if(nextNWP.valid())
		{
		// 	comm.addCircle(nextNWP.x, nextNWP.y, 50);
		// 	comm.addLine(nextWP.x, nextWP.y, nextNWP.x, nextNWP.y);
		}

		float motionAngle = Vector2D<int>::angle(nextWP, botPos);

		float finalSlope;   // final slope the current bot motion should aim for!
		if(nextNWP.valid())
			finalSlope = Vector2D<int>::angle(nextNWP, nextWP);
		else
			finalSlope = Vector2D<int>::angle(nextWP, botPos);

		finalSlope=point_ball_angle;
		float turnAngleLeft = normalizeAngle(finalSlope - state.homePos[botID].theta); // Angle left to turn

		float omega = turnAngleLeft * MAX_BOT_OMEGA / (2 * PI); // Speedup turn
		if(omega < MIN_BOT_OMEGA && omega > -MIN_BOT_OMEGA)
		{
			if(omega < 0) omega = -MIN_BOT_OMEGA;
			else omega = MIN_BOT_OMEGA;
		}

		float dist = Vector2D<int>::dist(nextWP, botPos);  // Distance of next waypoint from the bot
		float theta =  motionAngle - state.homePos[botID].theta;               // Angle of dest with respect to bot's frame of reference

		float profileFactor = (dist * 2 / MAX_FIELD_DIST) * MAX_BOT_SPEED;
		if(profileFactor < MIN_BOT_SPEED)
			profileFactor = MIN_BOT_SPEED;
		else if(profileFactor > MAX_BOT_SPEED)
			profileFactor = MAX_BOT_SPEED;

		if(dist < BOT_POINT_THRESH)
		{
			if((turnAngleLeft) > -DRIBBLER_BALL_ANGLE_RANGE  && (turnAngleLeft) < DRIBBLER_BALL_ANGLE_RANGE)
			{
				return getRobotCommandMessage(botID, 0, 0, 0, 0, true); //kick
			}
			else{
				return getRobotCommandMessage(botID, 0, 0, omega, 0, true);
			}
		}
		else
		{
			return getRobotCommandMessage(botID, profileFactor * sin(-theta), profileFactor * cos(-theta), omega, 0, false);
		}
	}
}

File: src/skills_py/scripts/skills/sDefendPoint.py
import skill_node
import math
import sys

sys.path.insert(0, '../../../navigation_py/scripts/navigation')
sys.path.insert(0, '../../../navigation_py/scripts/navigation')
sys.path.insert(0, '../../../navigation_py/scripts/navigation/src')
sys.path.insert(0, '../../../plays_py/scripts/utils')

from config import *
from wrapperpy import *
from obstacle import Obstacle
from geometry import *

def execute(param, state, bot_id, pub):
    DEFEND_RADIUS = 50.0
    ob=Vector2D()
    botPos=Vector2D(int(state.homePos[bot_id].x), int(state.homePos[bot_id].y))
    point_ball_angle = math.atan2(state.ballPos.y-param.DefendPointP.y,state.ballPos.x-param.DefendPointP.x)
    dpoint =Vector2D(int(param.DefendPointP.x + (param.DefendPointP.radius)*math.cos(point_ball_angle)) , int(param.DefendPointP.y + (param.DefendPointP.radius)*math.sin(point_ball_angle)))
    obs = Vector_Obstacle()
    for i in range(0,len(state.homeDetected)):
        if state.homeDetected[i] and i != bot_id:
            o = Obstacle()     
            o.x=state.homePos[i].x
            o.y=state.homePos[i].y
            o.radius=3.3*BOT_RADIUS
            obs.push_back(o)

    for j in range(0,len(state.awayDetected)):
        if state.awayDetected[j]:
            o = Obstacle()
            o.x=state.awayPos[j].x
            o.y=state.awayPos[j].y
            o.radius=3.3*BOT_RADIUS
            obs.push_back(o)

    point=Vector2D()
    nextWP = Vector2D()
    nextNWP =Vector2D()

    # calling of pathplanner has to be noticed later

    pathPlanner=MergeSCurve() 
    pathPlanner.plan(botPos,dpoint,nextWP,nextNWP,obs,len(state.homeDetected) + len(state.awayDetected),bot_id,True)
    motionAngle = nextWP.angle(botPos)

    if nextNWP.valid() is True :            # this has to be checked 
        finalSlope = nextNWP.angle(nextWP)
    else :
        finalSlope = nextWP.angle(botPos)

    finalSlope=point_ball_angle

    turnAngleLeft = ob.normalizeAngle(finalSlope - state.homePos[bot_id].theta)
    omega = turnAngleLeft * MAX_BOT_OMEGA / (2 * math.pi)    #Speedup turn
    if omega < MIN_BOT_OMEGA and omega > -MIN_BOT_OMEGA :
        if omega < 0 :
            omega = -MIN_BOT_OMEGA
        else : 
            omega = MIN_BOT_OMEGA
    dist = nextWP.dist(botPos)               # Distance of next waypoint from the bot
    theta =  motionAngle - state.homePos[bot_id].theta   #Angle of dest with respect to bot's frame of reference
    profileFactor = (dist * 2 / MAX_FIELD_DIST) * MAX_BOT_SPEED

    if profileFactor < MIN_BOT_SPEED :
        profileFactor = MIN_BOT_SPEED
    elif profileFactor > MAX_BOT_SPEED :
		profileFactor = MAX_BOT_SPEED
    print "HEre!!"
    if dist < BOT_POINT_THRESH :
        if (turnAngleLeft) > -DRIBBLER_BALL_ANGLE_RANGE  and (turnAngleLeft) < DRIBBLER_BALL_ANGLE_RANGE :
            skill_node.send_command(pub, state.isteamyellow, bot_id, 0, 0, 0, 0, True) # previous comment here was kick , needs be checked 
        else :
            skill_node.send_command(pub, state.isteamyellow, bot_id, 0, 0, omega, 0, True)
    else :
        skill_node.send_command(pub, state.isteamyellow , bot_id,profileFactor * math.sin(-theta) , profileFactor * math.cos(-theta), omega, 0, False)


